File: fortran/life_bug.f90
program main
implicit none
    logical equal
    CHARACTER(1)::choice
    integer, parameter ::  rows = 3, columns = 3,gen=5
    integer, parameter ::  rows1 = 3, columns1 = 3,gen1=4
    double precision,parameter::some=0.619
    integer,DIMENSION(rows*columns)::now,outnow
    integer,DIMENSION(rows1*columns1)::now1,outnow1,final
    integer :: i
    call GET_COMMAND_ARGUMENT(1,choice)
    if (choice .EQ. 'a') then
        do i=1,rows*columns
            if (rand()<some) then 
                now(i) = 1
            else 
                now(i)=0
            endif
        end do
		call sleep(1)
        call cls()
        call hm()
        call life(now,rows,columns,gen,outnow)
    else
        now1=(/0,0,0,1,1,1,0,0,0/)
        call sleep(1)
        call cls()
        call hm()
        call life(now1,rows1,columns1,gen1,outnow1)
    endif
    if (choice .NE. 'a') then
       final = (/0,0,0,0,0,0,0,0,0/)
        equal = size(outnow1) == size(final)
        if (equal) then
            do i=1,size(outnow1)
                equal = equal .and. final(i) == outnow1(i) 
                if ( .not. equal ) then
                    print *, equal,final(i) ,outnow1(i)
                    exit
                endif
            enddo
        endif
        if(equal) then
            print *, "All good!!"
        else
            print *,"Not good!!"
        endif
    endif
end program main


subroutine life (now,rows,cols,gen,outnow)
implicit none     
    
    integer,INTENT(IN) :: rows,cols
    integer, INTENT(INOUT) :: gen
    integer,INTENT(INOUT) :: now(rows*cols)
    integer,INTENT(OUT) :: outnow(rows*cols)
    integer :: i, j 
    integer::d(1:SIZE(now,1))
    d=now
    write(*,*) "recursive ", gen
    do i = 1, rows*cols
        if (modulo(i-1,rows)==0) then
            write(*,*)
        endif
        if(d(i) .eq. 0) then
            write(*,"(1x, A)",advance="no") " "
        else
            write(*,"(1x, A)",advance="no") "o"
        endif
    end do
    write(*,*)
    call sleep(1)
    call cls()
    call hm()
    call live(now,rows,cols,gen,outnow)
    
end subroutine life


RECURSIVE subroutine live (a,r,c,gen,outnow)    
implicit none   
    integer,INTENT(IN):: r,c
    integer, INTENT(INOUT) ::a(r*c)
    integer, INTENT(OUT) ::outnow(r*c)
    integer, INTENT(INOUT) :: gen
    integer:: i,nei,m
    integer::b(1:SIZE(a,1))
     do i = 1,c*r
        nei=0
        b(i)=0
        if((i-1) .GE. 1) then
            nei= nei+ a(i-1)
        endif
        if((i+1) .LE. (r*c)) then
            nei= nei+a(i+1)
        endif
        if((i-r) .GE. 1) then
            nei= nei+a(i-r)
        endif
        if((i+r) .LE. (r*c)) then
            nei= nei+a(i+r)
        endif
        if((i-r-1) .GE. 1) then
            nei= nei+a(i-r-1)
        endif
        if((i-r+1) .GE. 1) then
            nei= nei+a(i-r+1)
        endif
        if((i+r-1) .LE. (r*c)) then
            nei= nei+a(i+r-1)
        endif
        if((i+r+1) .LE. (r*c)) then
            nei= nei+a(i+r+1)
        endif
        if(a(i)==0) then
            if(nei==3) then
                b(i)=1
            endif
        else
            if(.NOT. (nei .LT. 2) .AND. .NOT. (nei .GT. 3)) then
                b(i)=a(i)
            endif
        endif
    end do
    m = (gen-1)
    write(*,*) "recursive ", m
    do i = 1, r*c
        if (modulo(i-1,r)==0) then
            write(*,*)
        endif
        if(b(i) .eq. 0) then
            write(*,"(1x, A)",advance="no") " "
        else
            write(*,"(1x, A)",advance="no") "o"
        endif
    end do
    write(*,*)
    if(m .NE. 0) then
        call sleep(1)
        call cls()
        call hm()
        outnow=b
        call live(b,r,c,m,outnow)
    endif
end subroutine live 



subroutine cls()
implicit none
    print *, achar(27)//"[2J"
end subroutine cls

subroutine hm()
implicit none
    print *, achar(27)//"[1;1H"
end subroutine hm

File: go/life_bug.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"time"
	"strconv"
)

var return_val int = 0

func cls(){
	if runtime.GOOS == "windows"{
		clearScreen := exec.Command("cmd", "/c", "cls")
		clearScreen.Stdout = os.Stdout
		clearScreen.Run()
	}else{
		clearScreen := exec.Command("clear")
		clearScreen.Stdout = os.Stdout
		clearScreen.Run()
	}

}

func verify_generation(gen int, now []int) bool{

	equal := true
	final := []int{0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0, 0,0,0,0,0}
	if gen == 4 {
		final[0] = 1
		final[1] = 1
		final[21] = 1
	}

	for i, v := range now {
		equal = (final[i]==v && equal)
	    if !equal {
			fmt.Printf("%#v %#v %#v\n", equal, final[i], v)
			break
	    }
	}
	fmt.Printf("Your generation is: %#v", strconv.FormatBool(equal))
	return equal

}

func get_neighbors(now []int, ind int, rows int, cols int) int{

	/*
	This is a helper method to check if the index to be used in now slice
	valid. If valid, return array value at provided index else return 0
	 */

	if ind >= 0 && ind < rows*cols{
		return now[ind]
	}
	return 0
}

func live(now []int, rows int, cols int, generations int, debug int) int{

	/*
	This method prints gof board after every iteration and then updates
	the gof board. Stops when generations becomes 0.
	 */

	if generations < 1{
		sleep()
		cls()
		fmt.Println("\nGeneration", generations)
		if debug == 1 {
			if generations!=5 && !verify_generation(generations, now){
				return_val=1
			}
		}
		fmt.Printf("\n")
		for c := 0; c < rows*cols; c++{
			if now[c] == 1{
				fmt.Printf("o")
			}else{
				fmt.Printf("-")
			}
			if (c+1)%rows == 0{
				fmt.Printf("\n")
			}
		}
		return return_val
	}else{
		sleep()
		cls()
		fmt.Println("\nGeneration", generations)

		if debug == 1 {
			if generations!=5 && !verify_generation(generations, now){
				return_val=1
			}
		}

		fmt.Printf("\n")

		for c := 0; c < rows*cols; c++{
			if now[c] == 1{
				fmt.Printf("o")
			}else{
				fmt.Printf("-")
			}
			if (c+1)%rows == 0{
				fmt.Printf("\n")
			}
		}

		newNow := make([]int, rows*cols)
		for c := 0; c < rows*cols; c++ {

			neighbours := get_neighbors(now, c-1, rows, cols) +
				          get_neighbors(now, c+1, rows, cols) +
				          get_neighbors(now, c-rows-1, rows, cols) +
				          get_neighbors(now, c-rows, rows, cols) +
				          get_neighbors(now, c-rows+1, rows, cols) +
						  get_neighbors(now, c+rows-1, rows, cols) +
						  get_neighbors(now, c+rows, rows, cols) +
						  get_neighbors(now, c+rows+1, rows, cols)

			newNow[c] = now[c]
			if now[c] == 0{
				if neighbours == 3 {
					newNow[c] = 1
				}else{
					newNow[c] = 0
				}
			}else{
				if now[c] == 1 && neighbours==2{
					newNow[c] = 1
				}else{
					newNow[c] = 0
				}
			}
		}
		generations = generations -1
		live(newNow, rows, cols, generations, debug)
		return return_val
	}
}

func life(rows int, cols int, some float64, optional ...int) []int{

	/*
	This method initialized the game of life board. And
	then calls the method live, which recursively calls
	itself till generations becomes zero.
	 */

	var now = make([]int, rows*cols)
	for c := 0; c < rows*cols; c++ {
		if rand.Float64() < some {
			now[c] = 1
		}else{
			now[c] = 0
		}
	}
	return now
}

func sleep(){
	// method to sleep the program for 100ms
	time.Sleep(100*time.Millisecond)
}

func initialize() int{
	// initialize the stdout
	// sleeps for 1 sec and then clear screen
	time.Sleep(1*time.Second)
	cls()
	return 0
}


func main(){

	args := os.Args[1:]

	if len(args) > 0{
		initialize()
		now := life(50, 20, 0.619)
		live(now, 50, 20, 200, 0)
	}else{
		now := []int{1,1,0,1,1, 1,0,1,1,1, 1,1,0,1,0, 1,0,1,1,1, 1,1,0,1,1}
		if live(now, 5, 5, 5, 1) == 0{
			fmt.Println("All Good")
		}else {
			fmt.Println("Not Good")
		}
	}


}
